src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,243 @@
+    private final BlobManager blobManager;
+    private final ZipFileProcessor zipFileProcessor;
+    private final DocumentProcessor documentProcessor;
+    private final EnvelopeProcessor envelopeProcessor;
+
+    public UploadEnvelopeDocumentsService(
+        // only process envelopes after cool down period
+        // can be removed once uploading feature is removed from main job
+        @Value("${scheduling.task.upload-documents.min-envelope-age}") long minimumEnvelopeAge,
+        EnvelopeRepository envelopeRepository,
+        BlobManager blobManager,
+        ZipFileProcessor zipFileProcessor,
+        DocumentProcessor documentProcessor,
+        EnvelopeProcessor envelopeProcessor
+    ) {
+        this.minimumEnvelopeAge = minimumEnvelopeAge;
+        this.envelopeRepository = envelopeRepository;
+        this.blobManager = blobManager;
+        this.zipFileProcessor = zipFileProcessor;
+        this.documentProcessor = documentProcessor;
+        this.envelopeProcessor = envelopeProcessor;
+    }
+
+    public void processEnvelopes() {
+        envelopeRepository
+            .findByStatus(CREATED)
+            .stream()
+            // can be moved to query instead. but it won't be needed after upload is removed from main task
+            .filter(envelope -> envelope.getCreatedAt().isBefore(now().minus(minimumEnvelopeAge, MINUTES)))
+            .collect(groupingBy(Envelope::getContainer))
+            .forEach(this::processByContainer);
+    }
+
+    private void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Processing envelopes in {} container. Envelopes found: {}", containerName, envelopes.size());
+
+        try {
+            CloudBlobContainer blobContainer = blobManager.getContainer(containerName);
+
+            envelopes.forEach(envelope -> {
+                Optional<CloudBlockBlob> blobClient = getCloudBlockBlob(
+                    blobContainer,
+                    envelope.getZipFileName(),
+                    envelope.getId()
+                );
+
+                blobClient
+                    .flatMap(client -> blobManager
+                        .acquireLease(client, containerName, envelope.getZipFileName())
+                        .map(voidLeaseId -> client)
+                    )
+                    .flatMap(client ->
+                        getBlobInputStream(client, containerName, envelope.getZipFileName(), envelope.getId())
+                    ).flatMap(inputStream ->
+                        processInputStream(inputStream, containerName, envelope.getZipFileName(), envelope.getId())
+                    ).map(result ->
+                        uploadParsedZipFileName(envelope, result.getPdfs())
+                    ).filter(isUploaded -> isUploaded).ifPresent(voidTrue ->
+                        envelopeProcessor.handleEvent(envelope, DOC_UPLOADED)
+                    );
+
+                blobClient.ifPresent(this::breakLease);
```

Comment:
I don't think representing this logic as a chain of transformations is the clearest way. Also, in case of an unhandled exception:
- one blob can block other blobs from being processed
- lease won't be broken

I propose that individual methods return specific values and throw exceptions instead of returning empty optionals. Then this code look like this:
```
CloudBlockBlob blobClient;

try {
    blobClient = getCloudBlockBlob(
        blobContainer,
        envelope.getZipFileName(),
        envelope.getId()
    );

    blobManager.acquireLease(blobClient, containerName, envelope.getZipFileName())
    BlobInptuStream inputStream =
        getBlobInputStream(blobClient, containerName, envelope.getZipFileName(), envelope.getId());
        
    ProcessResult result = processInputStream(inputStream, containerName, envelope.getZipFileName(), envelope.getId());
    boolean isUploaded = uploadParsedZipFileName(envelope, result.getPdfs())

    if (isUploaded) {
        envelopeProcessor.handleEvent(envelope, DOC_UPLOADED)
    }
} catch (Exception exc) {
    // log error saying we failed to upload this envelope
} finally {
    if (blobClient != null) {
        blobClient.breakLease();
    }
}
```

Reply:
Worth checking what other people think here

Reply:
unchained as suggested with tiny internal exception to handle logging messages the way it was before - keeping them separate instead of generic "i failed" message